nav2_util/src/node_utils.cpp

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_util/node_utils.hpp"
+#include <chrono>
+#include <string>
+#include <algorithm>
+#include <cctype>
+
+using std::chrono::high_resolution_clock;
+using std::to_string;
+using std::string;
+using std::replace_if;
+using std::isalnum;
+
+namespace nav2_util
+{
+
+string sanitizeNodeName(const string & potential_node_name)
+{
+  string node_name(potential_node_name);
+  // read this as `replace` characters in `node_name` `if` not alphanumeric.
+  // replace with '_'
+  replace_if(begin(node_name), end(node_name),
+    [](auto c) {return !isalnum(c);},
+    '_');
+  return node_name;
+}
+
+std::string timeToString(size_t len)
+{
+  string output(len, '0');  // prefill the string with zeros
+  auto timepoint = high_resolution_clock::now();
+  auto timecount = timepoint.time_since_epoch().count();
+  auto timestring = to_string(timecount);
+  if (timestring.length() >= len) {
+    // if `timestring` is shorter, put it at the end of `output`
+    output.replace(0, len,
+      timestring,
+      timestring.length() - len, len);
+  } else {
+    // if `output` is shorter, just copy in the end of `timestring`
+    output.replace(len - timestring.length(), timestring.length(),
+      timestring,
+      0, timestring.length());
+  }
+  return output;
+}
+
+std::string generateInternalNodeName(const std::string & prefix)
+{
+  return sanitizeNodeName(prefix) + "_" + timeToString(8);
+}
+
+rclcpp::Node::SharedPtr generateInternalNode(const std::string & prefix)
```

Comment:
Is this intended to be code generally available for the implementation to use or is it support code for testing (or both perhaps)? If it's testing-specific code, should we have a place for such code or is util the right place? I'm wondering if it's worth separating.

Reply:
I was thinking it could be used anywhere we need to create a node internal to a class without bothering the client about that detail. So it would get used by ServiceClient, maybe DynamicParamsClient, and a few other places. It wasn't meant to be just for test code.

With some of the talk on the logger exception thread about nodes have unique names, I was afraid that these classes could end up creating nodes with identical names. I figured the solution to that was to add a randomized string to the end of the name, and I was trying to encapsulate the algorithm for that in this function.